src/seedu/addressbook/data/person/ContactDetail.java

```diff
@@ -0,0 +1,84 @@
+package seedu.addressbook.data.person;
+
+import seedu.addressbook.data.exception.IllegalValueException;
+
+/**
+ * Represents a person's contact detail
+ * Guarantees: immutable; is valid as declared in {@link #isValid(String)}
+ */
+public abstract class ContactDetail {
+    /**
+     * 
+     */
+    private String value;
+
+    /**
+     * Instance's value attribute will be set to String supplied to constructor
+     * @param value
+     * @throws IllegalValueException Thrown if value is invalid, error message implemented by inherited class
+     */
+    public ContactDetail(String value) throws IllegalValueException {
+        this.value = value.trim();
+        if (!isValid(value)) {
+            throw new IllegalValueException(getValidationErrorMessage());
+        }
+    }
+
+    public void printClass() {
+        System.out.print(this.getClass());
+    }
+
+    public static void main(String[] args) {
+        try {
+            new Email("valid@e.mail", false).printClass();
+            ;
+        } catch (IllegalValueException e) {
+            // TODO Auto-generated catch block
+            e.printStackTrace();
+        }
+    }
+
+    /**
+     * Returns error message if validation fails
+     * @return
+     */
+
+    abstract String getValidationErrorMessage();
+
+    /**
+     * Returns true if a given string is a valid value
+     */
+    public boolean isValid(String test) {
+        return test.matches(getValidationRegex());
```

Comment:
I don't quite get what you mean. Because getValidationRegex() is an abstract method a child class is forced to provide a regex. See [Address](https://raw.githubusercontent.com/funnieguy/addressbook-level2/8d892304d92aa04f3c2b1c0931d17f543f3b501a/src/seedu/addressbook/data/person/Address.java) where the child class doesn't need to worry about the checking of value since its done in constructor of super class.